fsgrid.hpp

```diff
@@ -47,72 +49,182 @@ struct FsGridTools{
+                        // << ". Is " << 10 * processBox[0] * processBox[1] * processBox[2] <<" a const?\n";
+
+
             }
          }
+      }
 
-         if(optimValue == std::numeric_limits<double>::max() ||
-               processDomainDecomposition[0] * processDomainDecomposition[1] * processDomainDecomposition[2] != nProcs) {
+      if(optimValue == std::numeric_limits<double>::max() ||
+            processDomainDecomposition[0] * processDomainDecomposition[1] * processDomainDecomposition[2] != nProcs) {
+         std::cerr << "FSGrid domain decomposition failed, are you running on a prime number of tasks?" << std::endl;
+         throw std::runtime_error("FSGrid computeDomainDecomposition failed");
+      }
+      std::cerr << "Legacy done "<< processDomainDecomposition[0] << " " << processDomainDecomposition[1] << " " << processDomainDecomposition[2] << " \n\n\n\n";
+   }
+      
+   //! Helper function to optimize decomposition of this grid over the given number of tasks
+   static void computeDomainDecomposition(const std::array<uint64_t, 3>& GlobalSize, uint64_t nProcs, std::array<int,3>& processDomainDecomposition, int stencilSize=1, int verbose = 0) {
+      // if(legacy)
+      // {
+      //    computeLegacyDomainDecomposition(GlobalSize, nProcs, processDomainDecomposition, stencilSize);
+      //    return;
+      // }
+      int myRank, MPI_flag;
+      MPI_Initialized(&MPI_flag);
+      if(MPI_flag){
+         MPI_Comm_rank(MPI_COMM_WORLD, &myRank); // TODO allow for separate communicator
+      }
+      else{
+         myRank = 0;
+      }
+
+      std::array<uint64_t, 3> systemDim;
+      std::array<uint64_t, 3> processBox;
+      std::array<uint64_t, 3> minDomainSize;
+      int64_t optimValue = std::numeric_limits<int64_t>::max();
+      std::vector<std::pair<int64_t,std::array<uint64_t,3>>> scored_decompositions;
+      scored_decompositions.push_back(std::pair<int64_t,std::array<uint64_t,3>>(optimValue,{0,0,0}));
+      for(int i = 0; i < 3; i++) {
+         systemDim[i] = GlobalSize[i];
+         if(GlobalSize[i] == 1) {
+            // In 2D simulation domains, the "thin" dimension can be a single cell thick.
+            minDomainSize[i] = 1;
+         } else {
+            // Otherwise, it needs to be at least as large as our ghost
+            // stencil, so that ghost communication remains consistent.
+            minDomainSize[i] = stencilSize;
+         }
+      }
+      processDomainDecomposition = {1, 1, 1};
+      for (uint64_t i = 1; i <= std::min(nProcs, GlobalSize[0]/minDomainSize[0]); i++) {
+         for (uint64_t j = 1; j <= std::min(nProcs, GlobalSize[1]/minDomainSize[1]) ; j++) {
+            if( i * j  > nProcs ){
+               break;
+            }
+            uint64_t k = nProcs/(i*j);
+            // No need to optimize an incompatible DD, also checks for missing remainders
+            if( i * j * k != nProcs ) 
+               continue;
+
+            processBox[0] = systemDim[0]/i;
```

Comment:
I guess these lines should read:
```
processBox[0] = calcLocalSize(systemDim[0],i,0);
processBox[1] = calcLocalSize(systemDim[1],j,0);
processBox[2] = calcLocalSize(systemDim[2],k,0);
```
to account properly for remainders. (We should calculate communication based on the largest local domain)

Reply:
bump to this - now that we don't need to calcualte legacy stuff, I think we should indeed evaluate based on largest communication domain.

Reply:
Want the singleton discouragement also on now?